### PR TITLE
[Merged by Bors] - refactor: do not import old-style `have`/`suffices`/`replace` within mathlib

### PR DESCRIPTION
**Pull request by eric-wieser**
This syntax remains available downstream with `import Mathlib.Tactic`, but is no longer available in mathlib itself.

This follows on from #10640, which remove all current uses of this syntax.

By removing these imports, we prevent further regressions in mathlib,
and save reviewers from having to look out for this in review.

In future we could delete this syntax entirely, but this would harm downstream code, especially mathport output.

---

- [x] depends on: #10640 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by sgouezel** [comment]
I think it's really strange to have a tactic in mathlib which is not usable in mathlib. I'd rather remove the file completely, for coherence and predictability.

**Comment by eric-wieser** [comment]
I'd rather leave it in for now so as not to break anyone's teaching material.

> I think it's really strange to have a tactic in mathlib which is not usable in mathlib.

I don't think this is all that far away from our existing rule of "don't use `import Mathlib.Tactic` in mathlib".

**Comment by sgouezel** [comment]
(Scratch that, my tests were done on a previous commit in whic `Replace` wasn't disabled, and it imports `Tactic.Have`).

**Comment by sgouezel** [comment]
It turns out that in fact there were some bad uses of `have` in the library. I have started fixing things.

**Comment by eric-wieser** [comment]
I now see that #6964 recommends not spending manual effort here, and waiting for a syntax transformation tool instead.

Sorry for the false claim that we already didn't use this syntax anywhere!

**Comment by sgouezel** [comment]
Given the examples I've fixed (some of them very obfuscated), I don't think an automatic tool could do this. So I think I'm gonna fix this by hand instead of waiting, if you don't mind.

**Comment by sgouezel** [comment]
Speaking of obfuscation, this one is quite nice:
```
  suffices; exact (encode_lt_pair cf cg).imp (fun h => lt_trans h this) fun h => lt_trans h this
  change _; simp [encodeCode_eq, encodeCode]
```
I think Mario knows too well for his own good how everything works.

File: Mathlib/Computability/PartrecCode.lean [hunk 1]
```diff
@@ -211,14 +211,14 @@ theorem encode_lt_pair (cf cg) :
 
 theorem encode_lt_comp (cf cg) :
     encode cf < encode (comp cf cg) ∧ encode cg < encode (comp cf cg) := by
-  suffices; exact (encode_lt_pair cf cg).imp (fun h => lt_trans h this) fun h => lt_trans h this
```
Comment (eric-wieser) [inline Mathlib/Computability/PartrecCode.lean:214]: The literal translation would be
```lean
suffices ?_ by
  exact (encode_lt_pair cf cg).imp (fun h => lt_trans h this) fun h => lt_trans h this
```
right? I don't mind you removing @digama0's golf here, but it's reassuring to know that these gymnastics remain possible with the new syntax.

**Comment by grunweg** [comment]
The two remaining linter warnings shouldn't be hard (just about line length).

**Comment by grunweg** [comment]
I could use a hand fixing two remaining tests.

**Comment by sgouezel** [comment]
Now that everything is fixed (hopefully), I have extracted another PR keeping just the fixes to the library, and not touching the tactic files. Hopefully it shouldn't be controversial and can be merged quickly, and then we will discuss here what to do of the tactic files (disable them or remove them).

**Comment by eric-wieser** [comment]
I've rebased this to just include the import changes.

Even though this may not be where we want to end up in the longer term, I think we should merge this quickly to avoid any old-style `have`s sneaking back in to mathlib. We can then decide what to do with the syntax later.

**Comment by sgouezel** [comment]
The current version does not disable old style `have`, as `Tactic.Common` still imports `Tactic.Replace`, which imports `Tactic.Have`.

File: Mathlib/Tactic/Have.lean [hunk 2]
```diff
@@ -12,6 +12,9 @@ import Lean.Meta.Tactic.Assert
 
 This file extends the `have`, `let` and `suffices` tactics to allow the addition of hypotheses to
 the context without requiring their proofs to be provided immediately.
+
+As a style choice, this should not be used in mathlib; but is provided for downstream users who
+preferred the old style.
```
Comment (sgouezel) [inline Mathlib/Tactic/Have.lean:17]: Could you add the same comment to `Tactic.Replace`?

Comment (eric-wieser) [reply Mathlib/Tactic/Have.lean:17]: Done

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**